Modules/_winapi.c

```diff
@@ -774,12 +774,100 @@ gethandle(PyObject* obj, const char* name)
     return ret;
 }
 
+static PyObject* sortenvironmentkey(PyObject *module, PyObject *item) {
+    return _winapi_LCMapStringEx_impl(NULL, LOCALE_NAME_INVARIANT, LCMAP_UPPERCASE, item);
+}
+
+static PyMethodDef sortenvironmentkey_def = {
+    "sortenvironmentkey",
+    _PyCFunction_CAST(sortenvironmentkey),
+    METH_O,
+    ""
+};
+
+static PyObject *
+normalize_environment(PyObject* environment) {
+    PyObject *result, *keys, *keyfunc, *sort, *args, *kwargs;
+
+    keys = PyMapping_Keys(environment);
+    if (keys == NULL) {
+        return NULL;
+    }
+
+    keyfunc = PyCFunction_New(&sortenvironmentkey_def, NULL);
+    sort = PyObject_GetAttrString(keys, "sort");
+    args = PyTuple_New(0);
+    kwargs = PyDict_New();
+    PyDict_SetItemString(kwargs, "key", keyfunc);
+    if (PyObject_Call(sort, args, kwargs) == NULL) {
+        goto error;
+    }
+
+    result = PyDict_New();
+    wchar_t *prev_key_string = NULL;
+
+    for (int i=0; i<PyList_GET_SIZE(keys); i++) {
+        PyObject *key = PyList_GET_ITEM(keys, i);
+        PyObject* value = PyObject_GetItem(environment, key);
+
+        if (! PyUnicode_Check(key) || ! PyUnicode_Check(value)) {
+            PyErr_SetString(PyExc_TypeError,
+                "environment can only contain strings");
+            Py_DECREF(result);
+            result = NULL;
+            goto error;
+        }
+        if (PyUnicode_FindChar(key, '\0', 0, PyUnicode_GET_LENGTH(key), 1) != -1 ||
+            PyUnicode_FindChar(value, '\0', 0, PyUnicode_GET_LENGTH(value), 1) != -1)
+        {
+            PyErr_SetString(PyExc_ValueError, "embedded null character");
+            Py_DECREF(result);
+            result = NULL;
+            goto error;
+        }
+        /* Search from index 1 because on Windows starting '=' is allowed for
+           defining hidden environment variables. */
+        if (PyUnicode_GET_LENGTH(key) == 0 ||
+            PyUnicode_FindChar(key, '=', 1, PyUnicode_GET_LENGTH(key), 1) != -1)
+        {
+            PyErr_SetString(PyExc_ValueError, "illegal environment variable name");
+            Py_DECREF(result);
+            result = NULL;
+            goto error;
+        }
+
+        if (i == 0) {
+            prev_key_string = PyUnicode_AsWideCharString(key, NULL);
+            continue;
+        }
+
+        wchar_t *key_string = PyUnicode_AsWideCharString(key, NULL);
+        if (CompareStringOrdinal(prev_key_string, -1, key_string, -1, TRUE) == CSTR_EQUAL) {
+            continue;
+        }
+        PyObject_SetItem(result, key, value);
+        prev_key_string = key_string;
```

Comment:
The `prev_key_string` you overwrite here is not freed with `PyMem_Free`

Reply:
Yes, fixed.